source/disruption/traffic_reports_api.cpp

```diff
@@ -121,50 +122,61 @@ NetworkDisrupt& TrafficReport::find_or_create(const type::Network* network){
-        }
-        type::Indexes stop_areas;
-
+                                   const std::string& filter,
+                                   const std::vector<std::string>& forbidden_uris,
+                                   const type::Data& d,
+                                   const boost::posix_time::ptime now) {
+
+   for (auto idx : network_idx) {
+       const auto* network = d.pt_data->networks[idx];
+       std::string new_filter = "network.uri=" + network->uri;
+       if (!filter.empty()) {
+           new_filter += " and " + filter;
+       }
+
+       type::Indexes stop_points;
        try {
-            stop_areas = ptref::make_query(type::Type_e::StopArea, new_filter, forbidden_uris, d);
-        } catch (const ptref::parsing_error& parse_error) {
-            LOG4CPLUS_WARN(logger, "Disruption::add_stop_areas : Unable to parse filter "
-                                + parse_error.more);
-        } catch (const ptref::ptref_error& /*ptref_error*/) {
-           // that can arrive quite often if there is a filter, and
-           // it's quite normal. Imagine /line/metro1/traffic_reports
-           // for the network SNCF.
-        }
-        for (auto stop_area_idx: stop_areas) {
-            const auto* stop_area = d.pt_data->stop_areas[stop_area_idx];
-            auto v = stop_area->get_publishable_messages(now);
-            for (const auto* stop_point: stop_area->stop_point_list) {
-                auto vsp = stop_point->get_publishable_messages(now);
-                v.insert(v.end(), vsp.begin(), vsp.end());
-            }
-            if (!v.empty()) {
-                NetworkDisrupt& dist = this->find_or_create(network);
-                auto find_predicate = [&](const std::pair<const type::StopArea*, DisruptionSet>& item) {
-                    return item.first == stop_area;
-                };
-                auto it = boost::find_if(dist.stop_areas, find_predicate);
-                if (it == dist.stop_areas.end()) {
-                    dist.stop_areas.push_back(std::make_pair(stop_area, DisruptionSet(v.begin(), v.end())));
-                } else {
-                    it->second.insert(v.begin(), v.end());
-                }
-            }
-        }
-    }
+           stop_points = ptref::make_query(type::Type_e::StopPoint, new_filter, forbidden_uris, d);
+       } catch (const ptref::parsing_error& parse_error) {
+           LOG4CPLUS_WARN(logger, "Disruption::add_stop_points : Unable to parse filter "
+                               + parse_error.more);
+       } catch (const ptref::ptref_error& /*ptref_error*/) {
+          // that can arrive quite often if there is a filter, and
+          // it's quite normal. Imagine /line/metro1/traffic_reports
+          // for the network SNCF.
+       }
+
+       // build a map of messages per stop_area (iterate only on stop_points of the network)
+       std::map<const nt::StopArea*, std::vector<boost::shared_ptr<nt::disruption::Impact>>> sa_messages;
```

Comment:
Maybe you need a `std::map<const nt::StopArea*, std::set<boost::shared_ptr<nt::disruption::Impact>>>`